test/include/seqan3/test/sandboxed_path.hpp

```diff
@@ -0,0 +1,600 @@
+ * for at appropiate places, and some caveats.
+ *
+ *  Invariant:
+ *  - seqan3::sandboxed_path is always converted to an absolute path
+ *  - seqan3::sandboxed_path always points to a file or directory inside a given sandbox directory
+ *  - The sandbox directory is immutable during the life cycle of a sandboxed_path
+ *
+ * Caveat:
+ *  - Relative paths are not possible
+ *  - Some functions will leave the sandboxed environment.
+ *    - calling relative_path() leaves the environment of sandboxed_path.
+ */
+class sandboxed_path : public std::filesystem::path
+{
+private:
+    std::filesystem::path const sandbox_directory;
+
+public:
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+
+    /*!\brief Construction of a sandboxed_path.
+     * \param directory must be an absolute path.
+     *
+     * After construction, the sandboxed_path will point to `path`.
+     */
+    explicit sandboxed_path(std::filesystem::path directory)
+        : std::filesystem::path{directory}
+        , sandbox_directory{std::move(directory)}
+    {
+        normalise();
+        check_invariant();
+    }
+
+    /*!\brief Construction from a given sandbox directory and a path within the sandbox directory.
+     * \param sandbox_directory The absolute path to the sandbox directory.
+     * \param path The relative or absolute path that must be inside the sandbox directory.
+     */
+    explicit sandboxed_path(std::filesystem::path sandbox_directory, std::filesystem::path path)
+        : std::filesystem::path {std::move(path)}
+        , sandbox_directory     {std::move(sandbox_directory)}
+    {
+        normalise();
+        check_invariant();
+    }
+
+    sandboxed_path() = delete;                           //!< Deleted.
+    sandboxed_path(sandboxed_path const&) = default;     //!< Defaulted.
+    sandboxed_path(sandboxed_path&&) noexcept = default; //!< Defaulted.
+    ~sandboxed_path() = default;                         //!< Defaulted.
+
+    /*!\brief Replaces the path with a new path.
+     * \param  new_path The new path.
+     *
+     * This works the same way as std::filesystem::path::operator=
+     * and additionally checks the invariant.
+     *
+     * ### Exceptions
+     *
+     * Basic exception guarantee.
+     */
+    sandboxed_path & operator=(sandboxed_path const & new_path)
+    {
+        std::filesystem::path::operator=(new_path);
+        check_invariant();
+        return *this;
+    }
+
+    /*!\brief Replaces the path with a new path.
+     * \param  new_path The new path.
+     *
+     * This works the same as std::filesystem::path::operator=
+     * and additionally checks the invariant.
+     *
+     * ### Exceptions
+     *
+     * Basic exception guarantee
+     */
+    sandboxed_path & operator=(sandboxed_path && new_path)
+    {
+        std::filesystem::path::operator=(std::move(new_path));
+        check_invariant();
+        return *this;
+    }
```

Comment:
Wouldn't `=default;` be ok here because the `new_path` must have had checked its invariant at some point?

Reply:
This would only be possible if `this->sandbox_directory == new_path.sandbox_directory` was true. But `new_path` can have a path to a different sandbox. The way I designed this class is that `sandbox_directory` is unchangeable, even under a move.

Reply:
Ah, right :)